anchor/network/src/handshake/mod.rs

```diff
@@ -0,0 +1,252 @@
+        node_info.seal(&self.keypair).unwrap()
+    }
+
+    fn verify_node_info(&mut self, node_info: &NodeInfo) -> Result<(), Error> {
+        let ours = self.node_info_provider.get_node_info().network_id;
+        if node_info.network_id != *ours {
+            return Err(Error::NetworkMismatch {
+                ours,
+                theirs: node_info.network_id.clone(),
+            });
+        }
+        Ok(())
+    }
+
+    fn handle_handshake_request(
+        &mut self,
+        peer_id: PeerId,
+        request: Envelope,
+        channel: ResponseChannel<Envelope>,
+    ) {
+        // Handle incoming request: send response then verify
+        let response = self.sealed_node_record();
+        let _ = self.behaviour.send_response(channel, response.clone()); // Any error here is handled by the InboundFailure handler
+
+        self.unmarshall_and_verify(peer_id, &request);
+    }
+
+    fn handle_handshake_response(&mut self, peer_id: PeerId, response: &Envelope) {
+        self.unmarshall_and_verify(peer_id, response);
+    }
+
+    fn unmarshall_and_verify(&mut self, peer_id: PeerId, response: &Envelope) {
+        let mut their_info = NodeInfo::default();
+
+        if let Err(e) = their_info.unmarshal(&response.payload) {
+            self.events.push(Event::Failed {
+                peer_id,
+                error: Error::NodeInfo(e),
+            });
+        }
+
+        match self.verify_node_info(&their_info) {
+            Ok(_) => self.events.push(Event::Completed {
+                peer_id,
+                their_info,
+            }),
+            Err(e) => self.events.push(Event::Failed { peer_id, error: e }),
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler =
+        <RequestResponseBehaviour<Codec> as NetworkBehaviour>::ConnectionHandler;
+    type ToSwarm = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.behaviour.handle_established_inbound_connection(
+            connection_id,
+            peer,
+            local_addr,
+            remote_addr,
+        )
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        role_override: Endpoint,
+        port_use: PortUse,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.behaviour.handle_established_outbound_connection(
+            connection_id,
+            peer,
+            addr,
+            role_override,
+            port_use,
+        )
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        // Initiate handshake on new connection
+        if let FromSwarm::ConnectionEstablished(conn_est) = &event {
```

Comment:
There might be an issue here. We should initiate the handshake only for an outbound connection.

Reply:
This has been changed